### PR TITLE
Add first draft CTL implementation.

### PR DESCRIPTION
**Pull request by nick-shaw**

### CODE REVIEW

File: model/gamut_compress.ctl [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+
+// calculate compressed distance
+float compress(float dist, float lim, float thr, bool invert, float power)
+{
+    float cdist;
+    float s;
+    if (dist < thr) {
+        cdist = dist;
+    }
+    else { // power(p) compression function plot https://www.desmos.com/calculator/54aytu7hek
```
Comment (Tristimulus) [inline model/gamut_compress.ctl]: this should probably be removed
not sure how long desmos keeps these links active

Comment (carolalynn) [reply model/gamut_compress.ctl]: Agreed, we can host these links in the separate documentation instead.

File: model/gamut_compress.ctl [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+    // convert to ACEScg
+    float lin_AP1[3] = mult_f3_f44(ACES, AP0_2_AP1_MAT);
+
+    // thr is the percentage of the core gamut to protect.
+    float thr[3] = {
+        min(0.9999, threshold_c),
+        min(0.9999, threshold_m),
+        min(0.9999, threshold_y)
+    };
```
Comment (Tristimulus) [inline model/gamut_compress.ctl]: Since the threshold values are constants in this version there is no reason for this check.

Comment (nick-shaw) [reply model/gamut_compress.ctl]: Agreed. I have pushed a commit which removes code which is redundant when the parameters are constants.

File: model/gamut_compress.ctl [hunk 3]
```diff
@@ -0,0 +1,126 @@
+
+// <ACEStransformID>urn:ampas:aces:transformId:v1.5:LMT.Academy.GamutCompress.a1.0.1</ACEStransformID>
```
Comment (scottdyer) [inline model/gamut_compress.ctl]: Should TransformID and ACESuserName be here yet? I'd expect these in a pull request to the ampas/aces-dev repository but thinking maybe it's premature to have them in this repository?

```suggestion
// <ACEStransformID>urn:ampas:aces:transformId:v1.5:LMT.VWG.GamutCompress.a1.v1</ACEStransformID>
```

**Review by scottdyer — COMMENTED** [review]
To resolve the TransformID/ACESuserName comment, I'd suggest either removing them completely (for now) or replacing the Academy namespace in the TransformID with "VWG" and using the “vendor-supplied” format (until submission to the main aces-dev repository).

I suppose you could also keep them as-is and change the filename of the transform itself to `LMT.Academy.GamutCompress.ctl`, as if you were already prepping for a pull request to aces-dev. 

Also, if there is another commit before merging, perhaps conform all the comments to title case - but that's a minor issue.

Overall, looks good to me.

**Comment by thomasmansencal** [comment]
Just to be clear as discussions on Slack have shown it was not, here is what I'm suggesting:

From this:

```c
// Percentage of the core gamut to protect.
const float threshold_c = 0.815;
const float threshold_m = 0.803;
const float threshold_y = 0.880;

// Agressiveness of the compression curve
const float power = 1.2;

// Distance from achromatic which will be compressed to the gamut boundary
const float cyan =  1.147;
const float magenta = 1.264;
const float yellow = 1.312;

// calculate compressed distance
float compress(float dist, float lim, float thr, bool invert, float power)
```

To that:

```c
// Distance from achromatic which will be compressed to the gamut boundary
const float lim_cyan =  1.147;
const float lim_magenta = 1.264;
const float lim_yellow = 1.312;

// Percentage of the core gamut to protect.
const float thr_cyan = 0.815;
const float thr_magenta = 0.803;
const float thr_yellow = 0.880;

// Agressiveness of the compression curve
const float power = 1.2;

// calculate compressed distance
float compress(float dist, float lim, float thr, float power, bool invert)
```

**Comment by thomasmansencal** [comment]
Oh and comments either end-up with a punctuation mark or they don't since we are being pedantic :) I prefer with!

**Comment by carolalynn** [comment]
I am happy with all of these changes - thanks Nick et al! I don't really have a problem with the 'p' variable as it's just storage for repeated use.

**Comment by thomasmansencal** [comment]
As I was wondering if we could do something about better consistency in the case style, i.e. _camelCase_ vs _underscores_, I looked at `aces-dev` and got sad:

```c
    // OCES to RGB rendering space
    float rgbPre[3] = mult_f3_f44( oces, AP0_2_AP1_MAT);
    ...
    // Scale and clamp white to avoid casted highlights due to D60 simulation
    linearCV[0] = min( linearCV[0], 1.0) * SCALE;
    linearCV[1] = min( linearCV[1], 1.0) * SCALE;
    linearCV[2] = min( linearCV[2], 1.0) * SCALE;
    ...
    // Apply gamma adjustment to compensate for dim surround
    linearCV = darkSurround_to_dimSurround( linearCV);
    // Default output is full range, check if legalRange param was set to true
    if (legalRange == 1) {
        outputCV = fullRange_to_smpteRange_f3( outputCV);
    }
```

Local variables tend to be _camelCase_ though.

**Comment by thomasmansencal** [comment]
I noted that there is an empty line at the top of the file! To go back to my previous comment, maybe an actual consistent naming scheme would be like that:

```c
// <ACEStransformID>urn:ampas:aces:transformId:v1.5:LMT.VWG.GamutCompress.a1.v1</ACEStransformID>
// <ACESuserName>ACES 1.3 Look - Gamut Compress</ACESuserName>

//
// "LMT" for compressing out of gamut scene-referred values into AP1.
// 
// Input and output are ACES2065-1.
//

//
// Direction:
//    By default this transform operates in the forward direction (compressing the gamut).
//    If instead an inverse operation is needed (undoing a prior gamut compression), there
//    is a runtime flag to do this. In ctlrender, this can be achieved by appending 
//    '-param1 invert 1' after the '-ctl gamut_compress.ctl' string.
//



import "ACESlib.Transform_Common";



/* --- Gamut Compress Parameters --- */
// Distance from achromatic which will be compressed to the gamut boundary
const float LIM_CYAN =  1.147;
const float LIM_MAGENTA = 1.264;
const float LIM_YELLOW = 1.312;

// Percentage of the core gamut to protect
const float THR_CYAN = 0.815;
const float THR_MAGENTA = 0.803;
const float THR_YELLOW = 0.880;

// Agressiveness of the compression curve
const float PWR = 1.2;



// Calculate compressed distance
float compress(float dist, float lim, float thr, float pwr, bool invert)
{
    float comprDist;
    float scl;
    float nd;
    float p;

    if (dist < thr) {
        comprDist = dist; // No compression below threshold
    }
    else {
        // Calculate scale factor for y = 1 intersect
        scl = (lim - thr) / pow(pow((1.0 - thr) / (lim - thr), -pwr) - 1.0, 1.0 / pwr);

        // Normalize distance outside threshold by scale factor
        nd = (dist - thr) / scl;
        p = pow(nd, pwr);

        if (!invert) {
            comprDist = thr + scl * nd / (pow(1.0 + p, 1.0 / pwr)); // Compress
        }
        else {
            if (dist > (thr + scl)) {
                comprDist = dist; // Avoid singularity
            }
            else {
                comprDist = thr + scl * pow(-(p / (p - 1.0)), 1.0 / pwr); // Uncompress
            }
        }
    }

    return comprDist;
}



void main 
(
    input varying float rIn, 
    input varying float gIn, 
    input varying float bIn, 
    input varying float aIn,
    output varying float rOut,
    output varying float gOut,
    output varying float bOut,
    output varying float aOut,
    input uniform bool invert = false
) 
{ 
    // Source values
    float ACES[3] = {rIn, gIn, bIn};

    // Convert to ACEScg
    float linAP1[3] = mult_f3_f44(ACES, AP0_2_AP1_MAT);

    // Achromatic axis
    float ach = max_f3(linAP1);

    // Distance from the achromatic axis for each color component aka inverse RGB ratios
    float dist[3];
    if (ach == 0.0) {
        dist[0] = 0.0;
        dist[1] = 0.0;
        dist[2] = 0.0;
    }
    else {
        dist[0] = (ach - linAP1[0]) / fabs(ach);
        dist[1] = (ach - linAP1[1]) / fabs(ach);
        dist[2] = (ach - linAP1[2]) / fabs(ach);
    }

    // Compress distance with parameterized shaper function
    float comprDist[3] = {
        compress(dist[0], LIM_CYAN, THR_CYAN, PWR, invert),
        compress(dist[1], LIM_MAGENTA, THR_MAGENTA, PWR, invert),
        compress(dist[2], LIM_YELLOW, THR_YELLOW, PWR, invert)
    };

    // Recalculate RGB from compressed distance and achromatic
    float comprLinAP1[3] = {
        ach - comprDist[0] * fabs(ach),
        ach - comprDist[1] * fabs(ach),
        ach - comprDist[2] * fabs(ach)
    };

    // Convert back to ACES2065-1
    ACES = mult_f3_f44(comprLinAP1, AP1_2_AP0_MAT);

    // Write output
    rOut = ACES[0];
    gOut = ACES[1];
    bOut = ACES[2];
    aOut = aIn;
}
```

**Comment by nick-shaw** [comment]
All (well most) of the CTL files in `aces-dev` have a blank line at the top, so I just copied that as a template. Maybe it wasn't deliberate.

**Pull request merged**